collect_app/src/main/java/org/odk/collect/android/widgets/PrinterWidget.kt

```diff
@@ -0,0 +1,134 @@
+import org.odk.collect.android.formentry.questions.QuestionDetails
+import org.odk.collect.android.utilities.QuestionMediaManager
+import org.odk.collect.androidshared.livedata.MutableNonNullLiveData
+import org.odk.collect.androidshared.livedata.NonNullLiveData
+import org.odk.collect.async.Scheduler
+import org.odk.collect.material.MaterialProgressDialogFragment
+import org.odk.collect.printer.HtmlPrinter
+import org.odk.collect.qrcode.QRCodeCreator
+import java.io.ByteArrayOutputStream
+
+class PrinterWidget(
+    context: Context,
+    questionDetails: QuestionDetails,
+    private val viewModel: PrinterWidgetViewModel,
+    private val scheduler: Scheduler,
+    private val questionMediaManager: QuestionMediaManager,
+    private val qrCodeCreator: QRCodeCreator,
+    private val htmlPrinter: HtmlPrinter
+) : QuestionWidget(context, questionDetails) {
+
+    init {
+        render()
+
+        MaterialProgressDialogFragment.showOn(
+            context as AppCompatActivity,
+            viewModel.isLoading,
+            context.supportFragmentManager
+        ) {
+            MaterialProgressDialogFragment().also { dialog ->
+                dialog.message = context.getString(org.odk.collect.strings.R.string.loading)
+            }
+        }
+    }
+
+    override fun onCreateAnswerView(context: Context, prompt: FormEntryPrompt, answerFontSize: Int): View {
+        val answerView = LayoutInflater.from(context).inflate(R.layout.printer_widget, null)
+        answerView
+            .findViewById<MaterialButton>(R.id.printer_button)
+            .setOnClickListener {
+                print()
+            }
+        return answerView
+    }
+
+    override fun setOnLongClickListener(listener: OnLongClickListener?) = Unit
+
+    override fun getAnswer(): IAnswerData? = null
+
+    override fun clearAnswer() = Unit
+
+    override fun registerToClearAnswerOnLongPress(activity: Activity?, viewGroup: ViewGroup?) = Unit
+
+    private fun print() {
+        formEntryPrompt.answerText?.let {
+            viewModel.parseAndPrint(scheduler, it, questionMediaManager, qrCodeCreator, context, htmlPrinter)
+        }
+    }
+}
+
+class PrinterWidgetViewModel : ViewModel() {
+    private val _isLoading = MutableNonNullLiveData(false)
+    val isLoading: NonNullLiveData<Boolean> = _isLoading
+
+    fun parseAndPrint(
+        scheduler: Scheduler,
+        htmlDocument: String,
+        questionMediaManager: QuestionMediaManager,
+        qrCodeCreator: QRCodeCreator,
+        context: Context,
+        htmlPrinter: HtmlPrinter
+    ) {
+        scheduler.immediate(
+            background = {
+                _isLoading.postValue(true)
+
+                val document = Jsoup.parse(htmlDocument)
+
+                parseImages(document, questionMediaManager)
+                parseQRCodes(document, qrCodeCreator)
+
+                document.html()
+            },
+            foreground = { content ->
+                htmlPrinter.print(context, content)
```

Comment:
Could this happen in the background as well?

Reply:
No, it would throw an exception:
`java.lang.RuntimeException: WebView cannot be initialized on a thread that has no Looper.`